### PR TITLE
[eslint] strip tailing property in assignments

### LINKED ISSUE #15510: [ESLint] Assignment like foo.bar.baz = X should warn about foo.bar instead
```js
  let foo = {}
  useEffect(() => {
    foo.bar.baz = 43;
  }, []);
```

This asks you to include `foo.bar.baz` into deps. But this doesn't make sense, as you write to it. Instead it should ask to include `foo.bar` into array.

If `foo` is a newly created object in render scope then we might additionally nudge you to `useRef` for it instead. Because otherwise it won't be shared between renders.

### PR DESCRIPTION
**Pull request by Zzzen**
closes #15510

>``` 
>let foo = {}
>useEffect(() => {
>   foo.bar.baz = 43;
>}, []);
>```
>This asks you to include foo.bar.baz into deps. But this doesn't make sense, as you write to it. Instead it should ask to include foo.bar into array.

### CODE REVIEW

**Comment by gaearon** [comment]
Can you please rebase? We moved to the suggestions API so the test needs tweaks.

**Comment by Zzzen** [comment]
👌 I'll try this weekend.

**Comment by Zzzen** [comment]
Moved to the suggestions API. PTAL

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js [hunk 1]
```diff
@@ -1437,6 +1437,25 @@ function getDependency(node) {
     )
   ) {
     return getDependency(node.parent);
+  } else {
+    return stripTailingPropInAssignment(node);
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1441]: What makes you think it's an assignment in this branch? Can we make it more specific and throw on something unexpected?

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js [hunk 2]
```diff
@@ -1437,6 +1437,25 @@ function getDependency(node) {
...
+}
+
+/**
+ * Assuming () means the passed/returned node:
+ * (props) => (props)
+ * (props.foo) => (props.foo)
+ * (props.foo.bar) = X => (props.foo).bar = X
+ * (props.foo.bar.baz) = X => (props.foo.bar).baz
+ */
+function stripTailingPropInAssignment(node) {
```
Comment (gaearon) [inline packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js:1452]: Just inline in this function into the parent one.

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js (hunk 1 above)
Comment (Zzzen) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js]: 😅  `tryStrippingTrailingPropsInAssigment` could have been a better choice. I'm trying to enumerate all cases and confused by a test:

```js
function MyComponent(props) {
  useEffect(() => {
    if (props.foo.onChange) {
      props.foo.onChange();
    }
  }, []);
}
```

Why is `props.foo` a proper dependency? Why not `props.foo.onChange`?

File: packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js (hunk 2 above)
Comment (Zzzen) [reply packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js]: 👌

**Comment by gaearon** [comment]
Thanks.

**Pull request merged**